lib/Mojo/mysql/PubSub.pm

```diff
@@ -100,14 +131,57 @@ sub _wait_db {
     my $sync_db = $self->mysql->db;
     return (delete $self->{wait_db}, $self->_wait_db($sync_db)) if $err;
     $res->finish;
-    $db->query('select sleep(600)', $cb);
-    $sync_db->query('update mojo_pubsub_subscribe set ts = current_timestamp where pid = ?', $db->pid);
+    _query_with_retry(
+      $db,
+      'select sleep(600)',
+      $cb
+    );
+    _query_with_retry(
+      $sync_db,
+      'update mojo_pubsub_subscribe set ts = current_timestamp where pid = ?',
+      $db->pid
+    );
     $self->_notifications($self->mysql->db);
   };
 
   warn qq|[PubSub] (@{[$wait_db->pid]}) reconnect\n| if DEBUG;
   $self->emit(reconnect => $wait_db);
-  return $wait_db->query('select sleep(600)', $cb);
+  return _query_with_retry(
+    $wait_db,
+    'select sleep(600)',
+    $cb
+  );
+}
+
+sub _query_with_retry {
+  my ($db, @query_args) = @_;
+
+  my $wantarray = wantarray;
+  my @results;
+
+  my $remaining_attempts = 2;
+  while ($remaining_attempts--) {
+    local $@;
+    eval {
+      if    (!defined $wantarray) {               $db->query(@query_args) }
+      elsif          ($wantarray) { @results    = $db->query(@query_args) }
+      else                        { $results[0] = $db->query(@query_args) }
+    };
+    last unless $@;  # success
+    die unless $remaining_attempts;  # rethrow $@ if no remaining attempts
+
+    # If we are allowed to retry, check if the error message looks
+    # like it refers to something retryable.  Only look within the
+    # first line to avoid potential spurious matches if the error
+    # e.g. contains a stack trace.
+    my $err = $@;                                  # avoid stringifying $@ ...
+    die unless $err =~ /^\V*(?:retry|timeout)/i;   # ... and maybe rethrow it
+
+    # If we got here, we are retrying the query:
+    warn qq|[PubSub] (@{[$db->pid]}) Retrying after: !!! $err\n| if DEBUG;
```

Comment:
Not sure how useful this is: Does it say anything about what it will retry? How about changing to something like:

```
my ($self, $sql, @bind) = @_; # instead of my ($db, @query_args) = @_;
...
warn qq|[PubSub] (@{[$db->pid]}) retry ($sql) !!! $err\n|
```

Reply:
I think the error includes the SQL but I can't find a copy of the error text right now.  Probably can't hurt to explicitly include the SQL in the warning though.